portfolio/src/main/java/com/google/sps/servlets/GalleryImagesServlet.java

```diff
@@ -0,0 +1,139 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+
+import com.google.gson.Gson;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.net.MalformedURLException;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Map;
+
+/** Servlet that returns gallery image information.*/
+@WebServlet("/gallery-images")
+public class GalleryImagesServlet extends HttpServlet {
+  private ArrayList<String> images = new ArrayList<>();
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    loadStoredGalleryImages();
+
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(images));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Get the message entered by the user.
+    String message = request.getParameter("message");
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(request, "image");
+
+    // Store image link and data in Datastore
+    storeGalleryImage(imageUrl);
+
+    response.sendRedirect("/gallery.html");
+  }  
+
+  /**
+   * Helper function that loads the class's ArrayList 
+   * with `numComments` of the comments in Datastore.
+   * All comments will be loaded if numComments < 0 or null.
```

Comment:
It doesn't look like this function deals with determining the number of comments to load. If it happens elsewhere in you application, I would remove this aspect of the comment because it will be hard to remember to update if the logic ever changes.

Reply:
Right 🤦 Dangers of copying over code. I've fixed the comment to reflect retrieval of images, not comments.